src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1343 @@
+
+  make_string_function(
+    goto_program,
+    i_it,
+    lhs,
+    new_function_type,
+    function_name,
+    new_args,
+    location,
+    signature);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_char_array_function_call
+
+  Inputs: a position in a goto program and a function name
+
+ Purpose: at the given position replace `r.some_function(arr,...)` by
+          `r=function_name({arr.length, arr.data}, ...)`
+
+\*******************************************************************/
+
+void string_refine_preprocesst::make_char_array_function_call(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const irep_idt &function_name,
+  const std::string &signature)
+{
+  make_char_array_function(
+    goto_program, i_it, function_name, signature, 1, true, true);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_char_array_side_effect
+
+  Inputs: a position in a goto program and a function name
+
+ Purpose: replace `r=s.some_function(i,arr,...)` by
+          `s=function_name(s,{arr.length,arr.data})`
+          and add a correspondance from r to s in the `string_builders` map
+
+\*******************************************************************/
+
+void string_refine_preprocesst::make_char_array_side_effect(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const irep_idt &function_name,
+  const std::string &signature)
+{
+  make_char_array_function(
+    goto_program, i_it, function_name, signature, 1, true, false);
+  code_function_callt &function_call=to_code_function_call(i_it->code);
+  string_builders[function_call.lhs()]=function_call.arguments()[0];
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::process_arguments
+
+  Inputs: a goto program, a position, a list of expressions, a location and a
+          signature
+
+ Outputs: a list of expressions
+
+ Purpose: for each expression that is a string or that is at a position with
+          an 'S' character in the signature, we declare a new `cprover_string`
+          whose contents is deduced from the expression and replace the
+          expression by this cprover_string in the output list;
+          in the other case the expression is kept as is for the output list.
+
+\*******************************************************************/
+
+exprt::operandst string_refine_preprocesst::process_arguments(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const exprt::operandst &arguments,
+  const source_locationt &location,
+  const std::string &signature)
+{
+  exprt::operandst new_arguments;
+
+  for(unsigned i=0; i<arguments.size(); i++)
+  {
+    exprt arg=arguments[i];
+    auto it=string_builders.find(arg);
+    if(it!=string_builders.end())
+      new_arguments.push_back(it->second);
+    else
```

Comment:
Is it intentional that the `else` block does not insert into `string_builders` (well, maybe some function call does)?